### ISSUE TITLE
async block with futures::stream::Buffered is not Send

### ISSUE THREAD

**Issue by mikeyhew** [rust-lang/rust#104382]
This issue was originally reported in https://github.com/rust-lang/futures-rs/issues/2636, but it looks like a bug in the compiler.

Here is the example from that issue (posted by @Tuetuopay):

```rust
use futures::stream::{empty, StreamExt};
use std::future::ready;

fn send<T: Send>(_: T) {}

fn main() {
    send(async {
        empty().map(ready::<&()>).buffered(0).next().await
    });
}
```

This should compile without issues, however rustc outputs this error message:

```
error: higher-ranked lifetime error
 --> src/main.rs:7:5
  |
7 | /     send(async {
8 | |         empty().map(ready::<&()>).buffered(0).next().await
9 | |     });
  | |______^
  |
  = note: could not prove `impl futures::Future<Output = Option<&'a ()>>: std::marker::Send`
```

This occurs on current nightly (2022-11-11) and stable. Playground link: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=43484ffb5887c99e0845737c0241c853

There appears to be some issue in how the compiler checks the async block. If you call `.boxed()` on the stream before calling `.buffered()` on it, the error goes away:

```rust
use futures::stream::{empty, StreamExt};
use std::future::ready;

fn send<T: Send>(_: T) {}

fn main() {
    send(async {
        empty().map(ready::<&()>).boxed().buffered(0).next().await
    });
}
```

This compiles without errors on current nightly and stable. Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9c086487d1ff76b309796efa45846f67

**Comment by compiler-errors** [comment]
This comment isn't particularly helpful except for other contributors, but this is another manifestation of the common "futures have trouble with higher-ranked lifetimes and auto traits" issue that comes from lifetime erasure in generator interiors, I think...

**Comment by mikeyhew** [comment]
@compiler-errors out of curiosity are there any GitHub issues or Zulip chat histories worth reading that describe the general problem you're referring to?

**Comment by zirconium-n** [comment]
> @compiler-errors out of curiosity are there any GitHub issues or Zulip chat histories worth reading that describe the general problem you're referring to?

https://github.com/rust-lang/rust/issues/64552
https://github.com/rust-lang/rust/pull/92449

**Comment by eholk** [comment]
Out of curiosity, does the issue still show up if you do `rustc -Z drop-tracking`?

I would guess yes since drop tracking doesn't do much with lifetime issues, but it'd still be interesting to know.

**Comment by khuey** [comment]
`-Z drop-tracking` crashes the compiler.

`thread '<unnamed>' panicked at 'type variables should not be hashed: _#11t', /rustc/b7bc90fea3b441234a84b49fdafeb75815eebbab/compiler/rustc_
type_ir/src/lib.rs:679:17`

**Comment by compiler-errors** [comment]
@khuey I've got a fix for that ICE coming up.

**Comment by vincenzopalazzo** [comment]
Now, this issue can be close with https://github.com/rust-lang/rust/pull/104753 ?

cc @compiler-errors

**Comment by compiler-errors** [comment]
No, #104753 only fixes this ICE (https://github.com/rust-lang/rust/issues/104382#issuecomment-1324410781) mentioned by khuey.

**Comment by jbg** [comment]
rustc also seems to have trouble identifying the location of the error. In a large application with a top-level async fn that branches out to call many others, if you introduce code similar to that in the first comment in a function deep within the application, the `error: higher-ranked lifetime error` points at the top-level fn (the 'root' of the 'tree of futures'), making debugging difficult.
